locator/tablets.cc

```diff
@@ -700,6 +722,71 @@ class tablet_effective_replication_map : public effective_replication_map {
             , _sharder(*_tmptr, table)
     { }
 
+    // Populate the `_tablet_to_dc_replication_factor_map` with replication factors for each data center.
+    // For each tablet index in the tablet map, it retrieves the current replica set using `get_endpoints_for_reading` 
+    // and counts replicas per data center.  Each data center's replication factor (RF) is saved under 
+    // its index according to `_datacenter_map`.
+    seastar::future<> build_dc_replication_factor_map() {
+        const tablet_map& tablets = get_tablet_map();
+        const topology& topo = get_topology();
+        fill_datacenter_map();
+        _tablet_to_dc_replication_factor_map.reserve(tablets.tablet_count());
+        for (size_t tablet_index = 0; tablet_index < tablets.tablet_count(); ++tablet_index) {
+            const tablet_replica_set& replicas = get_endpoints_for_reading(tablet_id{tablet_index});
+            _tablet_to_dc_replication_factor_map.emplace_back();
+            auto& token_rf_list = _tablet_to_dc_replication_factor_map.back();
+            token_rf_list.resize(_datacenter_map.size(), 0);
+            for (const auto& node : replicas) {
+                const locator::node* node_ptr = topo.find_node(node.host);
+                if (node_ptr == nullptr) {
+                    continue;
+                }
+                const sstring& datacenter = topo.get_datacenter(node.host);
+                const auto it = _datacenter_map.find(datacenter);
+                if (it == _datacenter_map.end()) {
+                    tablet_logger.debug("Could not find datacenter: {} for node: {} ", datacenter, node);
+                    continue;
+                }
+                const auto index = it->second;
+                assert(index < token_rf_list.size());
+                ++token_rf_list[index];
+            }
+            co_await seastar::maybe_yield();
+        }
+        tablet_logger.info("Added: [{}] records to rf map", _tablet_to_dc_replication_factor_map.size());
+        co_return;
+    }
+
+    [[nodiscard]] size_t get_replication_factor(const dht::token id, const seastar::sstring& datacenter) const override {
+        tablet_logger.trace("Get RF for token {}, dc {}", id, datacenter);
+        auto tablet = get_tablet_map().get_tablet_id(id);
+        assert(tablet.id < _tablet_to_dc_replication_factor_map.size());
```

Comment:
We prefer on_internal_error() to assert() to not cause unnecessary cluster-wide outages.